ote/src/clj/ote/transit_changes/detection.clj

```diff
@@ -29,15 +31,50 @@
         id-vector (vec (sort < id-map))]
     id-vector))
 
-(defn calculate-package-hashes-for-service [db service-id package-count]
-  (let [package-ids (get-gtfs-packages db service-id package-count)]
+(defn hash-recalculations
+  "List currently running hash-recalculations"
+  [db]
+  (let [calculations (specql/fetch db :gtfs/hash-recalculation
+                                   (specql/columns :gtfs/hash-recalculation)
+                                   {:gtfs/completed op/null?})]
+    (when (pos-int? (count calculations))
+       {:calculations calculations})))
+
+ (defn- start-hash-recalculation [db packets-total user]
+  (let [id (specql/insert! db :gtfs/hash-recalculation
+                  {:gtfs/started (java.sql.Timestamp. (System/currentTimeMillis))
+                   :gtfs/packets-ready 0
+                   :gtfs/packets-total packets-total
+                   :gtfs/completed nil
+                   :gtfs/created-by (:id user) })]
+    id))
+
+ (defn- update-hash-recalculation [db packages-ready id]
+  (specql/update! db :gtfs/hash-recalculation
+                  {:gtfs/packets-ready packages-ready}
+                 {:gtfs/recalculation-id id}))
+
+ (defn- stop-hash-recalculation [db id]
+  (specql/update! db :gtfs/hash-recalculation
+                 {:gtfs/completed (java.sql.Timestamp. (System/currentTimeMillis))}
+                 {:gtfs/recalculation-id id}))
+
+(defn calculate-package-hashes-for-service [db service-id package-count user]
+  (let [package-ids (get-gtfs-packages db service-id package-count)
+        ;; When given service has packages, mark calculation started
+        recalculation-id (when package-ids
+                           (:gtfs/recalculation-id (start-hash-recalculation db package-count user)))]
     (log/info "Found " (count package-ids) " For service " service-id)
+
     (doall
-      (for [package-id package-ids]
+      (dotimes [i (count package-ids)]
```

Comment:
Isn't the `doall` unnecessary here since it's followed by `dotimes`?

Reply:
ah. yes it is. dotimes is not lazy like for. Changed